### PR TITLE
feat(common): ngLet directive

### LINKED ISSUE #15280: Template Local Variables
[X] feature request

**Current behavior**

The new syntax is a great improvement!

```
<div *ngIf="userStream|async as user">...</div>
```

... And was discussed at length in #15020.

**Expected behavior**

The ability to capture the async pipe output and make it available as a template variable, is **much appreciated and very useful**. Unfortunately, this capability is currently bound closely to `*ngIf`; it is only available is a feature of `*ngIf`. (I am aware that there is something similar for `*ngFor`, unrelated to this feature request).

Here's what I request: that this capability be yanked apart from the conditional (`*ngIf`), and made available elsewhere.

I don't know what syntax would make the most sense. Perhaps:

```
<div *ngLet="userStream|async as user">...</div>
```

Or alternatively; it would be excellent to have a syntax to directly use any observable anywhere in a template, with the reactivity handled by Angular, and without worrying about potentially multiple async pipe instances needlessly subscribing many times. For example, imagine the following syntax:

```
<p>Hello, {{ userStream^.name }}</p>
```

**What is the motivation / use case for changing the behavior?**

I have a use case in front of me, the details which don't matter, where:

* I have a handful of observable streams with data to use in the template.
* Some of them occasionally have falsely values; but I still want to use them even if that is the case.
* Currently I'm doing some hackery to deal with the falsely problem, and nesting a big chunk of my template in a handful of nested `*ngIf`s which exists solely for the purpose of giving access to this ability to capture the async values to template variables.

### PR DESCRIPTION
**Pull request by ibarsi**
`ngLet` exposes the result of an expression that is stored as a variable to its child view.

Fixes #15280

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: #15280

## What is the new behavior?

`ngLet` evaluates the expression and, if stored in a variable, provides the result to its child view.

### Storing falsy result in a variable

`ngIf` supports the common use case of storing a value in a variable and exposing it to its child view.

This works well for most cases. However, if the value stored is falsy (ie. `false`, `0`, `''`, etc) the child view will not be rendered. This can be problematic if the stored value is required to be displayed, or provided as an argument to an output or DOM event.

`ngLet` works around this issue by always rendering the child view, regardless of the result of its expression.

### Semantic meaning

Even in the cases where `ngIf` works, some semantic meaning is lost. Reading the code doesn't communicate whether the auther intended contitional rendering or simply needed to expose the result of the expression to its child view.

`ngLet` explicitly communicates semantic meaning, illustrating that the intent of the author's code was to expose the result of the expression, not conditional rendering.

### Syntax

```html
<div *ngLet="condition as value">{{value}}</div>
```

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information

N/A

### CODE REVIEW

**Comment by sebek64** [comment]
This feature is tightly related to https://github.com/angular/angular/pull/25251, which is in some sense more general, but maybe a bit less readable. For me (as the author of the other pull request), both solutions are quite fine. The fact that these pull requests emerge indicates that this is a real deficiency that should be addressed somehow, by merging either one (or by proposing a completely different solution).

File: packages/common/src/directives/ng_let.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+@Directive({selector: '[ngLet]'})
+export class NgLet {
+  private _context: NgLetContext = new NgLetContext();
+
+  constructor(
+      private _viewContainer: ViewContainerRef, private _templateRef: TemplateRef<NgLetContext>, ) {
+  }
+
+  @Input()
+  set ngLet(condition: any) {
```
Comment (sandangel) [inline packages/common/src/directives/ng_let.ts]: should we name it condition?
we can absolutely use something like `*ngLet="data$ | async as data"` here.

File: packages/common/src/directives/ng_let.ts [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    this._context.$implicit = this._context.ngLet = condition;
+    this._viewContainer.clear();
+    this._viewContainer.createEmbeddedView(this._templateRef, this._context);
```
Comment (sandangel) [inline packages/common/src/directives/ng_let.ts]: why you put it here? Is that mean we will createEmbeddedView everytime input change?

I think we just need to `createEmbeddedView` once in the `ngOnInit` hook and only changes the value of property in the context object when `Input` changes.
```ts
  @Input()
  set ngLet(value: any) {
    this._context.$implicit = this._context.ngLet = value;
  }

  constructor(private _vcr: ViewContainerRef, private _templateRef: TemplateRef<NgLetContext>) {}

  ngOnInit() {
    this._vcr.createEmbeddedView(this._templateRef, this._context);
  }
```

**Review by sandangel — CHANGES_REQUESTED** [review]
I think you should add more tests for some use case like nested *ngLet, usage with `async` pipe...

File: packages/common/test/directives/ng_let.spec.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+    beforeEach(() => {
+      TestBed.configureTestingModule({
+        declarations: [TestComponent],
+        imports: [CommonModule],
+      });
+    });
+
+    it('should support binding to variable using let', async(() => {
+         const template = '<span *ngLet="booleanCondition let v">{{v}}</span>';
```
Comment (sandangel) [inline packages/common/test/directives/ng_let.spec.ts]: Correct me if I'm wrong but I think the syntax should be `*ngLet="booleanCondition; let v"`

File: packages/common/src/directives/ng_let.ts (hunk 1 above)
Comment (ibarsi) [reply packages/common/src/directives/ng_let.ts]: Absolutely! It's not expected to be a "condition" at all. Thanks for pointing that out. Maybe `value` or `data` is more appropriate?

File: packages/common/src/directives/ng_let.ts (hunk 2 above)
Comment (ibarsi) [reply packages/common/src/directives/ng_let.ts]: This is an excellent point... I didn't even consider that updating `context` on input changes would be enough 🤔

This makes a lot of sense, will update 👍🏼

File: packages/common/test/directives/ng_let.spec.ts (hunk 3 above)
Comment (ibarsi) [reply packages/common/test/directives/ng_let.spec.ts]: It's good either way, the semicolon isn't required 🙂

**Comment by sandangel** [comment]
Hi, I think you should have a look at [`*ngIf` directive test cases](https://github.com/angular/angular/blob/master/packages/common/test/directives/ng_if_spec.ts). some tests I think you should have are binding in `ng-template`, nested `*ngLet`, update multiple nodes...

File: packages/common/test/directives/ng_let.spec.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    afterEach(() => { fixture = null !; });
+
+    beforeEach(() => {
+      TestBed.configureTestingModule({
+        declarations: [TestComponent],
+        imports: [CommonModule],
+      });
+    });
+
+    it('should support binding to observable using as', async(() => {
```
Comment (jotatoledo) [inline packages/common/test/directives/ng_let.spec.ts]: Dont use `EventEmitter` for the purpose of an `Observable` instance. Use a Subject or similar.

**Comment by IgorMinar** [comment]
hi there. thanks for the PR. I've posted on https://github.com/angular/angular/issues/15280#issuecomment-430479166 about how we should go about implementing this feature. The implementation in this PR can't be accepted as it would (eventually) result in perf issues. Thanks for exploring this problem space. Your work will inform future implementations.

**Pull request closed without merging**